sw/device/silicon_creator/mask_rom/mask_rom.c

```diff
@@ -50,169 +54,121 @@ static inline rom_error_t mask_rom_irq_error(void) {
-    // Module
-    if (true) {
-      // Unlock execution of ROM_EXT executable code (text) sections.
-      RETURN_IF_ERROR(epmp_state_check(&epmp));
-      mask_rom_epmp_unlock_rom_ext_rx(&epmp,
-                                      manifest_code_region_get(manifest));
-      RETURN_IF_ERROR(epmp_state_check(&epmp));
-
-      // Jump to ROM_EXT entry point.
-      uintptr_t entry_point = manifest_entry_point_get(manifest);
-      base_printf("rom_ext_entry: %p\r\n", entry_point);
-      ((romextimage_entry_point *)entry_point)();
-      // NOTE: never expecting a return, but if something were to go wrong
-      // in the real `jump` implementation, we need to enter a failure case.
-
-      // Boot Failure (no policy check)
-      // - Because we may have locked some write-enable bits that any following
-      //   manifest cannot change if it needs to, we have to just reboot here.
-      // boot_failed(boot_policy, current_rom_ext_manifest); // Boot Policy
-      // Module
-    }
-  }
+/**
+ * Boots a ROM_EXT.
+ *
+ * Note: This function should not return under normal conditions. Any returns
+ * from this function must result in shutdown.
+ *
+ * @param manifest Manifest of the ROM_EXT to boot.
+ * @return rom_error_t Result of the operation.
+ */
+static rom_error_t mask_rom_boot(const manifest_t *manifest) {
+  RETURN_IF_ERROR(keymgr_check_state(kKeymgrStateReset));
+  keymgr_set_next_stage_inputs(&manifest->binding_value,
+                               manifest->max_key_version);
+
+  // Unlock execution of ROM_EXT executable code (text) sections.
+  RETURN_IF_ERROR(epmp_state_check(&epmp));
+  mask_rom_epmp_unlock_rom_ext_rx(&epmp, manifest_code_region_get(manifest));
+  RETURN_IF_ERROR(epmp_state_check(&epmp));
+
+  // Jump to ROM_EXT entry point.
+  uintptr_t entry_point = manifest_entry_point_get(manifest);
+  base_printf("rom_ext_entry: %p\r\n", entry_point);
+  ((rom_ext_entry_point *)entry_point)();
 
   return kErrorMaskRomBootFailed;
 }
 
+/**
+ * Attempts to boot ROM_EXTs in the order given by the boot policy module.
+ *
+ * @return Result of the last attempt.
+ */
+static rom_error_t mask_rom_try_boot(void) {
+  boot_policy_manifests_t manifests = boot_policy_manifests_get();
+  rom_error_t error = kErrorMaskRomBootFailed;
+  for (size_t i = 0; i < ARRAYSIZE(manifests.ordered); ++i) {
+    error = mask_rom_verify(manifests.ordered[i]);
+    if (error != kErrorOk) {
+      continue;
+    }
+    // Boot fails if a verified ROM_EXT cannot be booted.
+    RETURN_IF_ERROR(mask_rom_boot(manifests.ordered[i]));
+    // `mask_rom_boot()` should never return `kErrorOk`, but if it does
+    // we must shut down the chip instead of trying the next ROM_EXT.
+    return kErrorMaskRomBootFailed;
+  }
+  return error;
+}
+
 void mask_rom_main(void) {
-  sec_mmio_init(shutdown_finalize);
-  rom_error_t error = mask_rom_boot();
+  mask_rom_init();
+
+  // FIXME: what (if anything) should we print at startup?
```

Comment:
We should link this as an issue, as this is a security concern.

Reply:
Created #7894 (also related to #5773).